design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,215 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Custom agent image controller
+
+## Status
+
+provisional
+
+## Summary
+
+Enable a custom image containing the ironic-python-agent to be built for each
+BareMetalHost, instead of requiring all hosts to use the same agent image. The
+interface to the image builder will be defined by a new RamDiskImageRequest CR
+in metal³, with users able to substitute their own implementations of image
+building (in most cases this is expected to involve light customisation of an
+existing image) by creating a controller. The baremetal-operator will create
+and use RamDiskImageRequests as required when configured to do so.
+
+## Motivation
+
+When running the ironic-python-agent (IPA) deploy ramdisk, the agent is reliant
+on information baked in to the ramdisk image or available to discover from the
+immediate environment in order to be able to contact Ironic. In most cases this
+involves DHCP supplying an IP address and default route.
+
+Some users would prefer not to run DHCP in their infrastructure, but rather use
+static IP addresses instead. In some cases, particularly when provisioning
+using virtualmedia and not using PXE, they might prefer that the network on
+which to contact Ironic is not an untagged VLAN. In many cases, servers will be
+connected to switches that use static link aggregators, and thus should set up
+NIC bonding to enable reliable two-way communication.
+
+Currently there is no way to use static IPs or configure any of the other
+options on a per-host basis because all hosts are booted using the same ramdisk
+image.
+
+### Goals
+
+- Allow each host to use its own deploy ramdisk image customised with an
+  appropriate network configuration.
+
+### Non-Goals
+
+- Allow users who expect to provision a host to specify their own custom image
+  containing IPA.
+- Use Ironic with images that don't contain IPA.
+- Implement custom provisioners not based on Ironic.
+
+## Proposal
+
+### User Stories
+
+As a data center operator, I would like to manage hosts with metal³ and deploy
+them via virtualmedia on networks where there is no DHCP-like configuration
+service.
+
+## Design Details
+
+A new Custom Resource type, `RamDiskImageRequest` will be created in the
+metal3.io group. The Spec will contain the following fields:
+
+- Name of metaData Secret
```

Comment:
To make it usable for CoreOS we need userData. MetaData uses the OpenStack format and is related to the instance, which may not exist at this point.

Reply:
I think we can drop the metaData secret - I had been thinking of it as we'll pass all of the stuff supplied by the cloud, as opposed to the stuff supplied by the user (userData), but that's the wrong way of thinking about it. We should pass the stuff that is not specific to a Nova instance, which is only the networkData.

As far as the CoreOS Ignition goes, I was expecting that we would build a fixed (not host-specific) Ignition into the image that knows how to interpret the network data (and loads the IPA container).